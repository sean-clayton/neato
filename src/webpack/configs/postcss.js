// @flow

import path from 'path'
import smartImport from 'postcss-smart-import'
import precss from 'precss'
import postCSSFlexbugsFixes from 'postcss-flexbugs-fixes'
import ExtractTextPlugin from 'extract-text-webpack-plugin'
import extensions from '../../extensions'

export default (neato: NeatoConfigType): Object => {
  type LoaderType = {
    loader: string,
    [queryParam: string]: any
  }

  const minifyOptions = neato.optimize
    ? {
      autoprefixer: {
        add: true,
        remove: true,
        browsers: ['last 2 version', 'ie >= 10']
      },
      discardComments: {
        removeAll: true
      },
      discardDuplicates: true,
      discardOverridden: true,
      colormin: true,
      convertValues: true,
      sourcemap: true
    }
    : { }

  const localCssLoaders: LoaderType[] = [
    {
      loader: 'css-loader',
      options: {
        modules: true,
        sourceMaps: true,
        importLoaders: 1,
        localIdentName: neato.optimize ? '[name]_[local]__[hash:base64:5]' : '[path][name]_[local]__[hash:base64:5]',
        ...minifyOptions
      }
    },
    {
      loader: 'postcss-loader'
    }
  ]

  const globalCssLoaders: LoaderType[] = [
    {
      loader: 'css-loader',
      options: {
        sourceMaps: true,
        ...minifyOptions
      }
    },
    {
      loader: 'postcss-loader'
    }
  ]

  const shouldExtract = neato.options.pages && neato.options.pages.length > 0 && neato.action === 'BUILD'

  return {
    postcss: (webpack: any): Object => ({
      plugins: [
        smartImport({
          root: neato.projectPath,
          path: [
            path.resolve(neato.projectPath, 'src'),
            path.resolve(neato.projectPath, 'node_modules')
          ],
          addDependencyTo: webpack
        }),
        precss,
        postCSSFlexbugsFixes
      ]
    }),

    module: {
      rules: [
        {
          test: extensions.test.styles,
          include: path.resolve(neato.projectPath, 'src'),
          exclude: path.resolve(neato.projectPath, 'src/styles'),
          use: shouldExtract
            ? ExtractTextPlugin.extract(localCssLoaders)
            : [ { loader: 'style-loader' }, ...localCssLoaders ]
        },
        {
          test: extensions.test.styles,
          include: [
            path.resolve(neato.projectPath, 'src/styles'),
            path.resolve(neato.projectPath, 'node_modules')
          ],
          use: shouldExtract
            ? ExtractTextPlugin.extract(globalCssLoaders)
            : [ { loader: 'style-loader' }, ...globalCssLoaders ]
        }
      ]
    },

    plugins: neato.action === 'BUILD'
      ? [new ExtractTextPlugin({
        filename: '[name]-[chunkhash].css',
        allChunks: true
      })]
      : []
  }
}
