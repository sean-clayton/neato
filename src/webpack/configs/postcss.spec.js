/* eslint-env jest */

import postcss from './postcss'

describe('postcss config', () => {
  it('Generates a postcss-loader config', () => {
    const neatoMock = {
      neatoPath: '/test/neato/path',
      projectPath: '/test/project/path',
      optimize: false,
      options: {
        pages: ['index', 'other-page']
      },
      action: 'DEV'
    }
    expect(postcss(neatoMock)).toMatchSnapshot()
  })
})
