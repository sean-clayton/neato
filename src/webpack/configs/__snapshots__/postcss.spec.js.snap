exports[`postcss config Generates a postcss-loader config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": "/test/project/path/src/styles",
        "include": "/test/project/path/src",
        "test": /(\.postcss|\.pcss|\.css)$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[path][name]_[local]__[hash:base64:5]",
              "modules": true,
              "sourceMaps": true,
            },
          },
          Object {
            "loader": "postcss-loader",
          },
        ],
      },
      Object {
        "include": Array [
          "/test/project/path/src/styles",
          "/test/project/path/node_modules",
        ],
        "test": /(\.postcss|\.pcss|\.css)$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMaps": true,
            },
          },
          Object {
            "loader": "postcss-loader",
          },
        ],
      },
    ],
  },
  "plugins": Array [],
  "postcss": [Function],
}
`;
